### PR TITLE
refactor(storage): recluster with segment cluster stat

### LINKED ISSUE #12153: feat: add cluster statistics for segment
```
pub struct Statistics {
    pub row_count: u64,
    pub block_count: u64,
    pub perfect_block_count: u64,

    pub uncompressed_byte_size: u64,
    pub compressed_byte_size: u64,
    pub index_size: u64,

    pub col_stats: HashMap<ColumnId, ColumnStatistics>,
    // add cluster_stats here
    pub cluster_stats: Option<ClusterStatistics>,
}
```

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Execute process:

1. read out a batch of segments (1000) .

2. based on the cluster stats of segment level, calculate and select the segments with the highest depth (limit max_threads * 4).

3. read the blocks, select the ones with the highest depth, and record the segments in which those blocks are located. 

4. do a recluster on the selected blocks to generate new blocks.

5. Sort the new blocks with the remaining blocks in the segments recorded in step 3, and generate new segments. 

6. In order to maintain the initial order as much as possible, replace the positions of the segments recorded in step 3 in the base snapshot with the newly generated segments. The extra newly generated segments are appended to the top.

7. commit

- Closes #12153

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12535)

### CODE REVIEW

**Comment by zhyass** [comment]
askbend:summary

**Comment by zhyass** [comment]
askbend:summary

**Comment by BohuTANG** [comment]
askbend:summary

**Comment by zhyass** [comment]
askbend:summary

File: src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs [hunk 1]
```diff
@@ -110,104 +134,210 @@ impl ReclusterMutator {
...
+            let default_cluster_key_id = Some(self.cluster_key_id);
+            selected_statistics.iter().for_each(|v| {
+                merge_statistics_mut(&mut self.removed_segment_summary, v, default_cluster_key_id)
+            });
         }
+        Ok(selected)
+    }
 
-        Ok(false)
+    pub fn select_segments(
```
Comment (JackTan25) [inline src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs:194]: An extreme situation, if a segment's min and max is same, should we skip it? just like block level = -1

File: src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs [hunk 2]
```diff
@@ -73,24 +81,40 @@ impl ReclusterMutator {
...
+        let max_memory_usage = self.ctx.get_settings().get_max_memory_usage()? as usize;
+        let memory_threshold =
+            cmp::min(mem_info.avail as usize * 1024, max_memory_usage) * 50 / 100;
+
+        let mut remained_blocks = Vec::new();
+        let mut selected = false;
         for (level, block_metas) in blocks_map.into_iter() {
-            if block_metas.len() < 2 {
+            let len = block_metas.len();
+            if selected || len < 2 {
```
Comment (JackTan25) [inline src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs:109]: I think this logic is an error, remain_blocks should be the blocks which is not selected for this recluster ? Maybe I'm wrong.

Comment (JackTan25) [reply src/query/storages/fuse/src/operations/mutation/recluster_mutator.rs:109]: I'm wrong, sorry

**Review by JackTan25 — APPROVED** [review]
LGTM. It's fine

**Comment by zhyass** [comment]
> tested with long-run script, 9000 iterations (both standalone and 3 nodes cluster, linux box + local fs).
> 
This situation is possible. Because recluster is leveled, it can only be guaranteed that the avg_depth of blocks at the same level is as expected.

**Review by dantengsky — APPROVED** [review]

**Comment by dantengsky** [comment]
please block the merge for a while,  let me test the commit [5282418](https://github.com/datafuselabs/databend/pull/12535/commits/5282418b6a76c0e80397f569c437b69e82cd7ec3)

**Pull request merged**